providers/resource/bin/wifi_interface_mode.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+import subprocess
+import re
+from typing import List, Tuple
+
+
+def run_command(command: List[str]) -> str:
+    try:
+        result = subprocess.run(
+            command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        return result.stdout.decode("utf-8")
+    except Exception as e:
+        raise SystemError("An unexpected error occurred: {}".format(e))
+
+
+def get_interfaces() -> List[Tuple[str, str]]:
+    output = run_command(["iw", "dev"])
+    device_pattern = r"phy#(\d+).*?Interface (\S+)"
+    if output:
+        matches = re.findall(device_pattern, output, re.DOTALL)
+        return [
+            (device_id, interface_name)
+            for device_id, interface_name in matches
+        ]
+    raise SystemExit(0)
+
+
+def get_wiphy_info() -> List[Tuple[str, List[str]]]:
+    # We can not use command "iw phy0 info" to get the infomation of sepcific
+    # interface.
+    # Because of "phy0" in command could be other pattern like "mwiphy0".
+    # And such pattern as "mwiphy0" will only show in the output of command
+    # "iw phy".
+    output = run_command(["iw", "phy"])
+    if output:
+        contents = re.split(r"Wiphy\s+", output)
+        index_pattern = r"phy(\d+)"
+        wiphy_info = []
+        for content in contents:
+            if content.strip():
+                match_index = re.search(
+                    index_pattern, content, re.DOTALL
+                ).group(1)
+                supported_mode_pattern = (
+                    r"Supported interface modes:\s*((?:\s*\*\s*[\w/-]+\s*)+)"
+                )
+                match_modes = re.search(
+                    supported_mode_pattern, content, re.DOTALL
+                )
+                if match_modes:
+                    modes = re.findall(r"\*\s*([\w/-]+)", match_modes.group(1))
+
+                    wiphy_info.append((match_index, modes))
+        return wiphy_info
+
+
+def print_supported_modes() -> str:
+    interfaces = get_interfaces()
+    wiphy_info = get_wiphy_info()
+    for wiphy_index, modes in wiphy_info:
+        for device_index, interface in interfaces:
+            if wiphy_index == device_index:
+                for mode in modes:
+                    print("{}_{}: supported".format(interface, mode))
```

Comment:
If you like the dict suggestion, consider the follwing:
```suggestion
def print_supported_modes() -> str:
    interfaces = get_interfaces()
    wiphy_info = get_wiphy_info()
    for wiphy_index, modes in wiphy_info.items():
        interface = interfaces[wiphy_index]
        for mode in modes:
            print(
                "{}_{}: supported".format(
                    interface, mode.replace("-", "_").replace("/", "_")
                )
            )
        print()
```

This is also the only major change I ask you to do, we are missing a print() here, if you have more than 1 wifi interface this produces a "broken" resource. Also note, you don't print the inteface name as an attribute, do you concatenate these via jinja somewhere? If so, don't, lets find a better solution. 

Note: some attributes of this resource I'm pretty sure you can't use in a resource expression, this is the output on my machine, `-` and `/` you can not use in a resource expression, so I'm not sure this does what you need.
```
wlan0_IBSS: supported
wlan0_managed: supported
wlan0_AP: supported
wlan0_AP/VLAN: supported
wlan0_monitor: supported
wlan0_P2P-client: supported
wlan0_P2P-GO: supported
wlan0_P2P-device: supported
```

A better resource would look like this in my opinion:
```
interface: wlan0
IBSS: supported
managed: supported
AP: supported
AP_VLAN: supported
monitor: supported
P2P_client: supported
P2P_GO: supported
P2P_device: supported
```

So that you can 
```
resource:
    (supported_modes.interface == {interface} and supported_modes.AP_VLAN == 'supported')
```

Reply:
If I change the output format of this resource job, that means the related jobs[1] has to revise.
Is that possible I can do something like this in requires field and it will go through whole section of resource job output in single job(e.g. It has more than one interfaces)
```
requires:
wifi_interface_mode.interface == {{interface}} and wifi_interface_mode.AP == 'supported'
```


[2] https://github.com/search?q=repo%3Acanonical%2Fcheckbox%20wifi_interface_mode&type=code

Reply:
I'm not sure I get your question but please note the `(` `)`, sadly the current resource expression semantic is broken if you don't put `()` around your statement but need an actual `and`

Reply:
What I mean is the related template job is use device to generate instead of this resource job. So will the `requires` go through whole resource jobs content?
e.g. I got two interfaces, so I will get resouce job out put like this.
```
interface: wlan0
IBSS: supported
managed: supported
AP: supported
AP_VLAN: supported
monitor: supported
P2P_client: supported
P2P_GO: supported
P2P_device: supported

interface: wlan1
IBSS: supported
managed: supported
AP: supported
AP_VLAN: supported
monitor: supported
P2P_client: supported
P2P_GO: supported
P2P_device: supported
```
And the template job generate by device will be like below. And can it able to work if that is the job of second interface? 
The template job for wlan1, will it be able to check the content of wlan1?
```
unit: template
template-resource: device
template-filter: device.category == 'WIRELESS' and device.interface != 'UNKNOWN'
template-engine: jinja2
template-unit: job
id: wireless/nmcli_wifi_ap_bg_{{ interface }}
template-id: wireless/nmcli_wifi_ap_bg_interface
_summary: Create 802.11b/g Wi-Fi AP on {{ interface }} using NetworkManager
_purpose: Create an 802.11b/g Wi-Fi access point on a specified interface using NetworkManager command-line tools.
command:
  net_driver_info.py "$NET_DRIVER_INFO"
  wifi_nmcli_test.py ap {{ interface }} bg
plugin: shell
category_id: wifi_ap
estimated_duration: 10
flags: preserve-locale also-after-suspend
requires:
 (wifi_interface_mode.interface == {{interface}} and wifi_interface_mode.AP == 'supported')
```

Reply:
If you put the `( )` around the expression, yes. I'm pretty sure. You can easily double check this but the expression 

`(wifi_interface_mode.interface == {{interface}} and wifi_interface_mode.AP == 'supported')` means
"There is a wifi_interface that is called {{interface}} and the same interface supports AP"

Without the parenthesis it doesnt right now. So do put them there and please check!